### PR TITLE
Resource: Git Integration

### LINKED ISSUE #1104: Add Dataverse Git Integration Resources
## User Story

As a **Terraform user**, I want to manage **Dataverse Git integration** through Terraform so that I can **declare and automate Azure DevOps repository bindings and solution branch bindings as part of my Power Platform ALM setup**.

Use case: platform teams want to provision a Dataverse environment, connect it to an Azure DevOps repository, and bind one or more unmanaged solutions to branches and folders in a repeatable Terraform workflow instead of relying on manual Maker UI steps.

## Resource

- **Resource Name:** `powerplatform_environment_git_integration`, `powerplatform_solution_git_branch`
- **Service Name:** `git_integration`
- **Documentation Link:**
  - [sourcecontrolconfiguration](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/sourcecontrolconfiguration?view=dataverse-latest)
  - [sourcecontrolbranchconfiguration](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/sourcecontrolbranchconfiguration?view=dataverse-latest)
  - [sourcecontrolcomponent](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/reference/sourcecontrolcomponent?view=dataverse-latest)

### Potential Terraform Configuration

```hcl
resource "powerplatform_environment_git_integration" "example" {
  environment_id    = powerplatform_environment.example.id
  git_provider      = "AzureDevOps"
  scope             = "Solution"
  organization_name = "example-org"
  project_name      = "example-project"
  repository_name   = "example-repo"
}

resource "powerplatform_solution_git_branch" "solution_one" {
  environment_id       = powerplatform_environment.example.id
  git_integration_id   = powerplatform_environment_git_integration.example.id
  solution_id          = powerplatform_solution.solution_one.id
  branch_name          = "main"
  upstream_branch_name = "main"
  root_folder_path     = "solutions/solution-one"
}

resource "powerplatform_solution_git_branch" "solution_two" {
  environment_id       = powerplatform_environment.example.id
  git_integration_id   = powerplatform_environment_git_integration.example.id
  solution_id          = powerplatform_solution.solution_two.id
  branch_name          = "main"
  upstream_branch_name = "main"
  root_folder_path     = "solutions/solution-two"
}

```

### Additional Validation Rules

- `environment_id` must reference an environment with Dataverse enabled.
- `git_provider` should support `AzureDevOps`.
- `scope` should support `Solution` and `Environment`.
- `project_name` is required for `AzureDevOps`.
- `organization_name`, `project_name`, `repository_name`, and `branch_name` should be validated against the Dataverse `gitorganizations`, `gitprojects`, `gitrepositories`, and `gitbranches` endpoints where applicable.
- `solution_id` must reference an existing unmanaged solution in the same environment.
- In `Solution` scope, solution bindings should only be allowed when the parent integration scope is `Solution`.
- In `Environment` scope, the environment-level resource should manage the root branch configuration and proactively enable eligible custom unmanaged solutions by setting `solution.enabledforsourcecontrolintegration = true`, while excluding platform-owned default solutions such as `Common Data Services Default Solution` and `Default Solution`.
- In `Environment` scope, the resource should not create per-solution `sourcecontrolbranchconfiguration` rows. Only the root branch row should exist.

### `powerplatform_environment_git_integration`

| Action | Verb | URL | Status Codes | Comments |
|--------|------|-----|--------------|----------|
| Create | PATCH | `/api/data/v9.0/organizations(<organization-id>)` | 204 | Sets `SourceControlIntegrationScope` via `orgdborgsettings` before creating records |
| Create | POST | `/api/data/v9.0/sourcecontrolconfigurations` | 204/201 | Creates parent repo binding |
| Create | POST | `/api/data/v9.0/sourcecontrolbranchconfigurations` | 204/201 | Creates root branch row for `partitionid=00000000-0000-0000-0000-000000000000` |
| Create/Update | PATCH | `/api/data/v9.0/solutions(<solution-id>)` | 204 | In `Environment` scope, proactively sets `enabledforsourcecontrolintegration=true` for eligible custom unmanaged solutions |
| Read | GET | `/api/data/v9.0/sourcecontrolconfigurations` | 200 | Read current environment repo binding |
| Read | GET | `/api/data/v9.0/sourcecontrolbranchconfigurations?partitionId=00000000-0000-0000-0000-000000000000` | 200 | Read root branch binding |
| Update | PATCH | `/api/data/v9.0/sourcecontrolconfigurations(<id>)` | 204 | Update repo metadata |
| Update | PATCH | `/api/data/v9.0/sourcecontrolbranchconfigurations(sourcecontrolbranchconfigurationid=<id>,partitionid='00000000-0000-0000-0000-000000000000')` | 204 | Update root branch binding |
| Delete | PATCH | `/api/data/v9.0/sourcecontrolbranchconfigurations(sourcecontrolbranchconfigurationid=<id>,partitionid='00000000-0000-0000-0000-000000000000')` | 204 | Set `statuscode=1` and poll until root row disappears |
| Delete | PATCH | `/api/data/v9.0/solutions(<solution-id>)` | 204 | In `Environment` scope, set `enabledforsourcecontrolintegration=false` for solutions that were proactively enabled by the resource |

### `powerplatform_solution_git_branch`

| Action | Verb | URL | Status Codes | Comments |
|--------|------|-----|--------------|----------|
| Create | POST | `/api/data/v9.0/sourcecontrolbranchconfigurations` | 204/201 | Creates solution partition binding |
| Read | GET | `/api/data/v9.0/sourcecontrolbranchconfigurations?partitionId=<solution-guid>` | 200 | UI uses `partitionId` lookup rather than direct row-id lookup |
| Update | PATCH | `/api/data/v9.0/sourcecontrolbranchconfigurations(sourcecontrolbranchconfigurationid=<id>,partitionid='<solution-guid>')` | 204 | Update branch, folder, and upstream branch |
| Delete | PATCH | `/api/data/v9.0/sourcecontrolbranchconfigurations(sourcecontrolbranchconfigurationid=<id>,partitionid='<solution-guid>')` | 204 | Set `statuscode=1` and poll until query returns `[]` |

### JSON

```json
{
  "gitprovider": 0,
  "organizationname": "example-org",
  "projectname": "example-project",
  "repositoryname": "example-repo"
}
```

```json
{
  "sourcecontrolconfigurationid@odata.bind": "/sourcecontrolconfigurations(00000000-0000-0000-0000-000000000001)",
  "partitionid": "11111111-1111-1111-1111-111111111111",
  "branchname": "main",
  "upstreambranchname": "main",
  "rootfolderpath": "solutions/example-solution"
}
```

## Definition of Done

- [x] Data Transfer Objects in `internal/services/git_integration/dto.go`
- [x] Resource Models in `internal/services/git_integration/models.go`
- [x] API Client functions in `internal/services/git_integration/api_git_integration.go`
- [x] Resource implementations in `internal/services/git_integration/resource_environment_git_integration.go` and `internal/services/git_integration/resource_solution_git_branch.go`
- [x] Unit tests in `internal/services/git_integration/resource_environment_git_integration_test.go` and `internal/services/git_integration/resource_solution_git_branch_test.go` for happy path, error conditions, and boundary cases
- [ ] Acceptance tests for happy path
- [x] Resources added to `internal/provider/provider.go` and `internal/provider/provider_test.go`
- [x] Examples in the `/examples/resources/powerplatform_environment_git_integration` and `/examples/resources/powerplatform_solution_git_branch` folders
- [x] Schema documented using `MarkdownDescription`
- [ ] Change log entry `changie new -k added`
- [x] Run `make precommit` before PR

Acceptance test note: CI-backed happy path coverage for these resources will require access to an Azure DevOps organization, project, repository, and branch that the Power Platform acceptance-test identity can see through the Dataverse Git selector endpoints.

### PR DESCRIPTION
**Pull request by AdamCoulterOz**
## Summary

This PR adds first-class Dataverse Git integration support to the provider for Azure DevOps-backed environments and solutions.

It introduces:
- `powerplatform_environment_git_integration`
- `powerplatform_solution_git_branch`

## Related Issue

Closes #1104

## What’s Included

- Added the new Git integration service under `internal/services/git_integration`
- Added `powerplatform_environment_git_integration`
- Added `powerplatform_solution_git_branch`
- Registered both resources in `provider.go` and `provider_test.go`
- Added unit coverage for lifecycle, validation, import, and edge cases
- Added examples and generated docs for both resources
- Added import support for both resources
- Narrowed scope to Azure DevOps only for now, since that is the path validated live

### `powerplatform_environment_git_integration`

This resource manages the environment-level Dataverse repository binding.

- uses `environment_id`
- supports `scope = "Solution"` and `scope = "Environment"`
- validates `organization_name`, `project_name`, and `repository_name` against Dataverse Git selector endpoints
- sets `SourceControlIntegrationScope` before creating the parent Git configuration, matching the Maker UI sequence

### `powerplatform_solution_git_branch`

This resource manages a solution-level Dataverse branch/folder binding.

- uses `solution_id` instead of exposing `partition_id`
- reads solution bindings by `partitionId`, matching the Maker UI / Web API behavior
- uses the Dataverse lifecycle shape observed from the UI for update/delete
- polls until the branch binding is actually removed from the `partitionId` query result

### `EnvironmentScope`

`EnvironmentScope` behavior is different from `SolutionScope`, and this PR mirrors the UI-backed behavior validated live:

- creates the parent `sourcecontrolconfiguration`
- creates the root `sourcecontrolbranchconfiguration` for `partitionId = 00000000-0000-0000-0000-000000000000`
- proactively enables eligible custom unmanaged solutions by setting `solution.enabledforsourcecontrolintegration = true`
- does **not** create per-solution branch configuration rows in this mode
- excludes built-in platform-owned default solutions from proactive enablement

## Validation

Has been run locally:
- `make precommit`

Has also been manually live-validated against real Dataverse environments for:
- `SolutionScope`
  - create environment repo binding
  - create multiple solution branch bindings
  - remove one solution binding while keeping siblings and parent intact
- `EnvironmentScope`
  - create environment repo binding
  - create root branch binding
  - proactively enable custom unmanaged solutions
  - tear down cleanly

## Acceptance Tests

CI-backed acceptance tests are not included in this PR.

Reason:
- the current upstream acceptance workflow provisions Power Platform test credentials
- these Git integration resources additionally need Azure DevOps fixtures that Dataverse can resolve through the Git selector endpoints
- to add true `TestAcc` coverage in CI, maintainers will need upstream-managed Azure DevOps org/project/repo/branch access for the acceptance-test identity

This can be added later once an upstream Azure DevOps acceptance fixture is available.

## Known Separate Follow-Ups

These are not introduced by this PR, but were encountered during live validation and are tracked separately:

- #1102
- #1103

### CODE REVIEW

File: internal/services/git_integration/validation.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+	environmentPrefix, solutionID, found := strings.Cut(configuredValue, "_")
+	if !found {
+		return "", fmt.Errorf("the `solution_id` value `%s` must be the `id` exported by `powerplatform_solution` for environment `%s`", configuredValue, environmentID)
+	}
+
+	if environmentPrefix != environmentID {
+		return "", fmt.Errorf("the `solution_id` value `%s` uses environment `%s`, but this resource is targeting environment `%s`", configuredValue, environmentPrefix, environmentID)
+	}
+
+	return solutionID, nil
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/validation.go]: `normalizeSolutionID` now rejects non-GUID environment and solution segments before they ever reach OData URLs.

File: internal/services/git_integration/resource_environment_git_integration.go [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.OneOf(gitProviderAzureDevOps),
+				},
+			},
+			"scope": schema.StringAttribute{
+				MarkdownDescription: "Source control integration scope for the environment. Use `Solution` for solution-level branch bindings and `Environment` for an environment-level binding. In `Environment` scope, the provider manages the root branch binding and proactively enables eligible visible unmanaged solutions in the environment while excluding platform-owned default solutions.",
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.OneOf(scopeEnvironment, scopeSolution),
```
Comment (mawasile) [inline internal/services/git_integration/resource_environment_git_integration.go]: const should be in caps and camel case

**Review by mawasile — CHANGES_REQUESTED** [review]
Don't we need data sources to read existing integrations?

**Comment by mawasile** [comment]
@copilot update user docs using make userdocs

File: internal/services/git_integration/validation.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/validation.go:185]: Addressed - `normalizeSolutionID` now `uuid.Parse`s both the environment prefix and the solution segment before anything is interpolated into OData filters/URLs, rejecting non-GUID input with a clear error, and compares the prefix to the configured `environment_id` case-insensitively in canonical form (locked by `validation_internal_test.go`).

File: internal/services/git_integration/api_git_integration.go [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+	}
+	if err := c.Api.HandleForbiddenResponse(resp); err != nil {
+		return err
+	}
+	if resp.HttpResponse.StatusCode == http.StatusNotFound {
+		return nil
+	}
+
+	return c.waitForSolutionGitBranchRemoval(ctx, environmentID, branchID, configurationID, partitionID)
+}
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration.go]: Addressed - `DeleteSolutionGitBranch` now performs the disable PATCH against the row actually discovered by the `(configurationID, partitionID)` lookup (`existingBranch.ID`) and feeds that same ID into the removal wait, so a recreated row with a drifted ID can no longer 404-loop until timeout (`TestUnitDeleteSolutionGitBranch_UsesLookedUpBranchID`).

File: internal/services/git_integration/resource_environment_git_integration.go [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+	if plan.Scope.ValueString() == scopeEnvironment {
+		if err := r.GitIntegrationClient.EnsureSolutionScopeRootBranch(ctx, plan.EnvironmentID.ValueString(), created.ID, created.OrganizationName, created.ProjectName, created.RepositoryName); err != nil {
+			resp.Diagnostics.AddError(fmt.Sprintf("Client error when creating the root Git binding for %s", r.FullTypeName()), err.Error())
+			return
+		}
+
+		if err := r.GitIntegrationClient.EnableEnvironmentScopeSolutions(ctx, plan.EnvironmentID.ValueString()); err != nil {
+			resp.Diagnostics.AddError(fmt.Sprintf("Client error when enabling environment-scoped solutions for %s", r.FullTypeName()), err.Error())
+			return
+		}
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_environment_git_integration.go:191]: Acknowledged as a real gap - Delete currently cannot revert `enabledforsourcecontrolintegration` on solutions that Environment scope enabled, because the changed set is not recorded. Fixing it properly needs private state (or an attribute) tracking which solutions the resource flipped, which is a design decision I want to settle rather than rush - this PR is back in draft partly for that reason. Deferred deliberately for now.

File: internal/services/git_integration/api_git_integration.go [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+	if err != nil {
+		return err
+	}
+
+	apiURL := helpers.BuildDataverseApiUrl(environmentHost, buildSourceControlBranchConfigurationCompositeKeyPath(branchID, partitionID), nil)
+	headers := http.Header{}
+	headers.Set("If-Match", "*")
+	resp, err := c.Api.Execute(ctx, nil, http.MethodPatch, apiURL, headers, disableSourceControlBranchConfigurationDto{
+		StatusCode: sourceControlBranchConfigurationStatusInactive,
+	}, []int{http.StatusNoContent, http.StatusNotFound, http.StatusForbidden}, nil)
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration.go]: Addressed together with the sibling thread - the disable PATCH and the removal wait both use the looked-up `existingBranch.ID` now.

File: internal/services/git_integration/validation.go [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+		return "", fmt.Errorf("the `solution_id` value `%s` must use a valid environment GUID prefix", configuredValue)
+	}
+
+	if _, err := uuid.Parse(solutionID); err != nil {
+		return "", fmt.Errorf("the `solution_id` value `%s` must use a valid Dataverse solution GUID suffix", configuredValue)
+	}
+
+	if environmentPrefix != environmentID {
+		return "", fmt.Errorf("the `solution_id` value `%s` uses environment `%s`, but this resource is targeting environment `%s`", configuredValue, environmentPrefix, environmentID)
+	}
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/validation.go]: Addressed - the comparison now parses both sides with `uuid.Parse` and compares canonical forms, so GUID case and formatting differences no longer reject valid ids.

File: internal/services/git_integration/api_git_integration.go [hunk 7]
```diff
@@ -0,0 +1,915 @@
...
+}
+
+func sourceControlIntegrationScopeFromOrgDbValue(value string) string {
+	switch strings.TrimSpace(value) {
+	case "EnvironmentScope":
+		return scopeEnvironment
+	case "SolutionScope":
+		return scopeSolution
+	default:
+		return ""
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration.go:872]: Addressed - an unknown or missing org-setting value now returns an explicit 'scope could not be determined' error instead of writing an empty string into the validated `scope` attribute (`TestUnitGetSourceControlIntegrationScope_RejectsUnknownScope`).

File: internal/services/git_integration/models.go [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+		projectName = types.StringValue(dto.ProjectName)
+	}
+
+	return EnvironmentGitIntegrationResourceModel{
+		ID:               types.StringValue(dto.ID),
+		EnvironmentID:    types.StringValue(environmentID),
+		GitProvider:      types.StringValue(gitProviderFromInt(dto.GitProvider)),
+		Scope:            types.StringValue(scope),
+		OrganizationName: types.StringValue(dto.OrganizationName),
+		ProjectName:      projectName,
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/models.go]: Addressed - organization/project/repository are now always written as concrete string values via `types.StringValue` on Read/Import, never null (`TestUnitConvertSourceControlConfigurationDtoToModel_UsesConcreteProjectNameValue`).

File: internal/services/git_integration/api_git_integration.go [hunk 9]
```diff
@@ -0,0 +1,920 @@
...
+}
+
+func (c *client) EnsureSolutionScopeRootBranch(ctx context.Context, environmentID, configurationID, organizationName, projectName, repositoryName string) error {
+	defaultBranch, err := c.GetGitRepositoryDefaultBranch(ctx, environmentID, organizationName, projectName, repositoryName)
+	if err != nil {
+		return err
+	}
+
+	existingBranch, err := c.lookupAnySolutionGitBranchByPartition(ctx, environmentID, configurationID, rootPartitionID)
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration.go]: Addressed - renamed to `EnsureRootBranchConfiguration` to reflect that both scopes use it.

File: internal/services/git_integration/models.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func gitProviderFromInt(value int) string {
+	switch value {
+	case 0:
+		return gitProviderAzureDevOps
+	default:
+		return ""
+	}
+}
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/models.go]: Resolved by removing the attribute entirely (ef1cd8bc, also requested in review): `git_provider` is gone from the schema and model since Azure DevOps is the only supported provider, so `gitProviderFromInt` no longer exists and nothing can write an empty provider into state.

File: examples/resources/powerplatform_environment_git_integration/resource.tf [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  dataverse = {
+    language_code     = "1033"
+    currency_code     = "USD"
+    security_group_id = var.security_group_id
+  }
+}
+
+resource "powerplatform_environment_git_integration" "example" {
+  environment_id    = powerplatform_environment.example.id
+  git_provider      = var.git_provider
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_environment_git_integration/resource.tf]: Done in ad804996 - both examples now use explicit literal values throughout (no `var.` references), so the registry docs render self-contained configs.

File: examples/resources/powerplatform_environment_git_integration/variables.tf [hunk 12]
```diff
@@ -0,0 +1,53 @@
+variable "environment_display_name" {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_environment_git_integration/variables.tf]: Done in ad804996 - values inlined and `variables.tf` removed from both example directories.

File: examples/resources/powerplatform_environment_git_integration/resource.tf [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+      source = "microsoft/power-platform"
+    }
+  }
+}
+
+provider "powerplatform" {
+  use_cli = true
+}
+
+# Known limitation: Dataverse Git integration currently works only with delegated
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_environment_git_integration/resource.tf:13]: Done in ad804996 - both resources are listed in the README preview section, each noting they require delegated user authentication (non-interactive identities such as service principals are not supported).

File: internal/services/git_integration/resource_environment_git_integration.go [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+				Required:            true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.RequiresReplace(),
+				},
+			},
+			"git_provider": schema.StringAttribute{
+				MarkdownDescription: "Git provider for the repository binding. Supported value is `AzureDevOps`.",
+				Required:            true,
+				Validators: []validator.String{
+					stringvalidator.OneOf(gitProviderAzureDevOps),
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_environment_git_integration.go]: Agreed and done in ef1cd8bc - `git_provider` is removed from the schema/model/docs/examples entirely. The wire payload is unchanged (the Azure DevOps enum value was always omitted via `omitempty`), and a test asserts configs setting the attribute now error. We can reintroduce it as an optional attribute if other providers ever ship.

File: internal/services/git_integration/resource_environment_git_integration.go [hunk 15]
```diff
@@ -0,0 +1,347 @@
...
+	r.ProviderTypeName = req.ProviderTypeName
+
+	ctx, exitContext := helpers.EnterRequestContext(ctx, r.TypeInfo, req)
+	defer exitContext()
+
+	resp.TypeName = r.FullTypeName()
+	tflog.Debug(ctx, fmt.Sprintf("METADATA: %s", resp.TypeName))
+}
+
+func (r *EnvironmentGitIntegrationResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_environment_git_integration.go:49]: Deliberately deferred for now - exposing branch and git-folder attributes changes the resource surface and interacts with the root-branch defaulting logic, so I want to design it properly rather than bolt it on; it is one of the two reasons this PR is back in draft. I will follow up with a proposal before implementing.

File: internal/services/git_integration/api_git_integration.go [hunk 16]
```diff
@@ -0,0 +1,920 @@
...
+	"github.com/microsoft/terraform-provider-power-platform/internal/api"
+	"github.com/microsoft/terraform-provider-power-platform/internal/constants"
+	"github.com/microsoft/terraform-provider-power-platform/internal/customerrors"
+	"github.com/microsoft/terraform-provider-power-platform/internal/helpers"
+	"github.com/microsoft/terraform-provider-power-platform/internal/services/environment"
+	"github.com/microsoft/terraform-provider-power-platform/internal/services/solution"
+)
+
+const (
+	sourceControlBranchConfigurationStatusActive   = 0
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration.go]: Done in ad804996 - the consts in `api_git_integration.go` are PascalCase now (`SourceControlBranchConfigurationStatusActive/Inactive`, `StabilizedReadCountRequired`), matching the style used elsewhere in the repo.

File: internal/services/git_integration/api_git_integration.go [hunk 17]
```diff
@@ -0,0 +1,920 @@
...
+	for {
+		configuration, err := c.GetEnvironmentGitIntegration(ctx, environmentID, configurationID)
+		if err != nil {
+			if !errors.Is(err, customerrors.ErrObjectNotFound) {
+				return nil, err
+			}
+			stableReads = 0
+		} else {
+			stableReads++
+			if stableReads >= stabilizedReadCountRequired {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration.go]: Because the Dataverse `sourcecontrolconfiguration` row is eventually consistent after the create POST: the row is not immediately readable, and during binding the service can create/recreate rows asynchronously. The create therefore polls until the configuration materializes, and readiness additionally requires `StabilizedReadCountRequired` (2) consecutive stable reads before we trust it - without that, an immediate read-back intermittently 404s or observes a row that is about to be replaced.

File: internal/services/git_integration/api_git_integration_internal_test.go [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+	httpmock.RegisterNoResponder(func(req *http.Request) (*http.Response, error) {
+		return nil, fmt.Errorf("no responder found for %s %s", req.Method, req.URL)
+	})
+
+	deleted := false
+
+	httpmock.RegisterRegexpResponder("GET", regexp.MustCompile(`^https://api\.bap\.microsoft\.com/providers/Microsoft\.BusinessAppPlatform/scopes/admin/environments/00000000-0000-0000-0000-000000000001\?%24expand=permissions%2Cproperties\.capacity%2Cproperties%2FbillingPolicy(%2Cproperties%2FcopilotPolicies)?&api-version=2023-06-01$`),
+		func(req *http.Request) (*http.Response, error) {
+			return httpmock.NewStringResponse(http.StatusOK, `{
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/api_git_integration_internal_test.go]: Done in ad804996 - all inline JSON in the internal API tests now loads from fixture files under `tests/api/` and `tests/shared/` via `httpmock.File`.

File: internal/services/git_integration/resource_environment_git_integration.go (hunk 2 above)
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_environment_git_integration.go:80]: Done in ad804996 - renamed to PascalCase along with the other consts (that file itself declares none; the ones it referenced live in `api_git_integration.go`).

File: internal/services/git_integration/resource_environment_git_integration_test.go [hunk 19]
```diff
@@ -0,0 +1,259 @@
...
+			}
+			if !strings.Contains(bodyText, "SourceControlIntegrationScope") {
+				return nil, fmt.Errorf("organization scope patch missing SourceControlIntegrationScope: %s", bodyText)
+			}
+			return httpmock.NewStringResponse(http.StatusNoContent, ""), nil
+		})
+
+	httpmock.RegisterRegexpResponder("PATCH", regexp.MustCompile(`^https://00000000-0000-0000-0000-000000000001\.crm4\.dynamics\.com/api/data/v9\.0/solutions(?:%28|\()(33333333-3333-3333-3333-333333333333|44444444-4444-4444-4444-444444444444)(?:%29|\))$`),
+		func(req *http.Request) (*http.Response, error) {
+			body, err := io.ReadAll(req.Body)
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_environment_git_integration_test.go:144]: Done in ad804996 - payloads moved to `.json` fixtures. Where a flow genuinely returns different states over time (create-then-read sequences), the responder keeps a counter but serves sequence-named fixture files (`get_sourcecontrolconfigurations_1.json`, `_2.json`, ...), the same pattern the repo's newer tests use - so updating payloads is still just editing JSON.

File: internal/services/git_integration/resource_environment_git_integration_test.go [hunk 20]
```diff
@@ -0,0 +1,259 @@
...
+			}
+			if !matched {
+				return nil, fmt.Errorf("unexpected solution enablement URL: %s", req.URL.String())
+			}
+			return httpmock.NewStringResponse(http.StatusNoContent, ""), nil
+		})
+
+	httpmock.RegisterRegexpResponder("GET", regexp.MustCompile(`^https://00000000-0000-0000-0000-000000000001\.crm4\.dynamics\.com/api/data/v9\.0/sourcecontrolconfigurations%28[0-9a-f-]{36}%29$`),
+		func(req *http.Request) (*http.Response, error) {
+			if configurationImplicitlyDeleted {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_environment_git_integration_test.go:172]: Same as above - fixtures with sequence-named files where the flow requires state progression (ad804996).

File: internal/services/git_integration/resource_solution_git_branch.go [hunk 21]
```diff
@@ -0,0 +1,294 @@
...
+	resp.TypeName = r.FullTypeName()
+	tflog.Debug(ctx, fmt.Sprintf("METADATA: %s", resp.TypeName))
+}
+
+func (r *SolutionGitBranchResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	ctx, exitContext := helpers.EnterRequestContext(ctx, r.TypeInfo, req)
+	defer exitContext()
+
+	resp.Schema = schema.Schema{
+		MarkdownDescription: "Manages a solution-level Dataverse Git branch binding. This maps to the documented `sourcecontrolbranchconfiguration` Dataverse table and links a solution partition to a branch and folder beneath an environment Git integration.\n\nKnown limitation: the underlying Power Platform Git integration bootstrap currently requires delegated user principal authentication with Azure DevOps access. Service principal, app-only, and OIDC pipeline identities are not currently supported by the backing Dataverse Git integration flow.",
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_solution_git_branch.go:52]: Done in ad804996 - `powerplatform_solution_git_branch` is in the README preview section alongside the environment resource.

File: examples/resources/powerplatform_solution_git_branch/import.sh [hunk 22]
```diff
@@ -0,0 +1,3 @@
+# Solution Git branch resources can be imported using environment_id/git_integration_id/solution_id
+# The final segment can be either the raw Dataverse solution id or the provider-formatted powerplatform_solution.id
+terraform import powerplatform_solution_git_branch.example 00000000-0000-0000-0000-000000000000/11111111-1111-1111-1111-111111111111/00000000-0000-0000-0000-000000000000_22222222-2222-2222-2222-222222222222
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_solution_git_branch/import.sh]: Done in ad804996 - the import example now uses the plain three-GUID `/`-separated form exactly as suggested (ImportState already accepted a raw solution GUID as the third segment).

File: examples/resources/powerplatform_solution_git_branch/resource.tf [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+
+provider "powerplatform" {
+  use_cli = true
+}
+
+# Known limitation: Dataverse Git integration currently works only with delegated
+# user principal authentication that also has Azure DevOps repository access.
+# Service principal, app-only, and OIDC pipeline identities are not supported.
+
+resource "local_file" "solution_settings_file" {
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_solution_git_branch/resource.tf]: Done in ad804996 - the example drops the solution zip import, the settings file and the count gymnastics; it is now a minimal two-resource config referencing a literal solution id.

File: examples/resources/powerplatform_solution_git_branch/resource.tf [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+  solution_file  = coalesce(var.solution_file, "${path.module}/../powerplatform_solution/TerraformTestSolution_Complex_1_1_0_0.zip")
+  settings_file  = local_file.solution_settings_file.filename
+}
+
+resource "powerplatform_environment_git_integration" "example" {
+  count = var.enable_git_binding ? 1 : 0
+
+  environment_id    = powerplatform_environment.example.id
+  git_provider      = var.git_provider
+  scope             = var.scope
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply examples/resources/powerplatform_solution_git_branch/resource.tf]: Done in ad804996 - values inlined here too.

File: internal/services/git_integration/resource_solution_git_branch_test.go [hunk 25]
```diff
@@ -0,0 +1,292 @@
...
+		})
+
+	httpmock.RegisterRegexpResponder("GET", regexp.MustCompile(`^https://00000000-0000-0000-0000-000000000001\.crm4\.dynamics\.com/api/data/v9\.0/sourcecontrolconfigurations%2811111111-1111-1111-1111-111111111111%29$`),
+		func(req *http.Request) (*http.Response, error) {
+			return httpmock.NewStringResponse(http.StatusOK, httpmock.File("tests/resource/environment_git_integration/get_sourcecontrolconfiguration_1.json").String()), nil
+		})
+
+	httpmock.RegisterResponder("POST", "https://00000000-0000-0000-0000-000000000001.crm4.dynamics.com/api/data/v9.0/sourcecontrolbranchconfigurations",
+		func(req *http.Request) (*http.Response, error) {
+			bodyBytes, err := io.ReadAll(req.Body)
```
[the remark this answers is not in the document]
Comment (AdamCoulterOz) [reply internal/services/git_integration/resource_solution_git_branch_test.go:62]: Done in ad804996 - same treatment: all payloads in `.json` fixture files, counters only where the flow requires sequenced states, each state its own file.

**Comment by AdamCoulterOz** [comment]
Parking until v5 as agreed — the git integration surface needs design maturation (attribute surface, revert-on-destroy semantics) and we're deliberately not rushing it into the current train. Branch stays live; will reopen against the v5 planning cycle.
